src/radical/pilot/resource_config.py

```diff
@@ -194,3 +201,546 @@ class ResourceConfig(ru.TypedDict):
 
 
 # ------------------------------------------------------------------------------
+#
+class ResourceOccupation(ru.TypedDict):
+
+    INDEX      = 'index'
+    OCCUPATION = 'occupation'
+
+    _schema = {
+        INDEX     : int,
+        OCCUPATION: float,
+    }
+
+    _defaults = {
+        INDEX     : 0,
+        OCCUPATION: FREE,
+    }
+
+    def __init__(self, from_dict: Optional[dict] = None,
+                       **kwargs) -> None:
+
+        super().__init__(from_dict=from_dict, **kwargs)
+
+
+    def __str__(self):
+        return self.__repr__()
+
+
+    def __repr__(self):

```

Comment:
I feel like this violates the principle of least astonishment.

As a caller to, e.g., `find_slots()` you get for example `Slot: {'cores': [0:1.00, 1:1.00, 2:1.00, 3:1.00], 'gpus': [0:1.00, 1:1.00], ...}`. 

This has 2 surprising effects: 
1) `slot['cores'][0]` is not equal to `slot.as_dict()['cores'][0]`, and 
2) a caller can call dict methods on what looks like a string, `slot['cores'][0]` returns for example `0:1.00` which one might not expect to have a `.keys()` method.

Is there a reason that the compact representation is needed?

Reply:
> I feel like this violates the principle of least astonishment.
> 
> As a caller to, e.g., `find_slots()` you get for example `Slot: {'cores': [0:1.00, 1:1.00, 2:1.00, 3:1.00], 'gpus': [0:1.00, 1:1.00], ...}`.
> 
> This has 2 surprising effects:
> 
>     1. `slot['cores'][0]` is not equal to `slot.as_dict()['cores'][0]`, and
> 
>     2. a caller can call dict methods on what looks like a string, `slot['cores'][0]` returns for example `0:1.00` which one might not expect to have a `.keys()` method.
> 
> 
> Is there a reason that the compact representation is needed?

You are right here.  We usually don't use `repr` much as serialization / deserialization is in json and msgpack, mostly, but that should not be a reason to be lazy here.  I'll fix that.